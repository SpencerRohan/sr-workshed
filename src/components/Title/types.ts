import { ObjectValues } from "@/types/object-values";
import { TITLE_HEADINGS, TITLE_VARIANTS } from "./constants";

export type TitleHeadings = ObjectValues<typeof TITLE_HEADINGS>;
export type TitleVariants = ObjectValues<typeof TITLE_VARIANTS>;
