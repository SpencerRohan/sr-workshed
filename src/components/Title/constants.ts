export const TITLE_HEADINGS = {
	H1: "h1",
	H2: "h2",
	H3: "h3",
	H4: "h4",
	H5: "h5",
	H6: "h6",
} as const;

export const TITLE_VARIANTS = {
	1: "1",
	2: "2",
	3: "3",
	4: "4",
	5: "5",
	6: "6",
} as const;
