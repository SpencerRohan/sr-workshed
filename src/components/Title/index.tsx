import { WithDataTestId } from "@/types/data-testid-attribute";
import cn from "@/utils/cn";
import { FC, HTMLProps, ReactNode } from "react";
import { TitleHeadings, TitleVariants } from "./types";

export type TitleProps = {
	children: ReactNode;
	as?: TitleHeadings; // Defaults to h1
	title?: TitleVariants; // Defaults to 1
} & WithDataTestId<HTMLProps<HTMLHeadingElement>>;

/**
 * Component: Title
 *
 * Primary means to create a HTML Heading Element.  This allows separation of
 * heading semantics and styles, which is key for accessibility and SEO.
 */
const Title: FC<TitleProps> = ({
	children,
	as: Tag = "h1",
	title = "1",
	...props
}: TitleProps) => {
	return (
		<Tag
			{...props}
			className={cn(
				{
					"text-t1-sm md:text-t1-md lg:text-t1-lg": title === "1",
					"text-t2": title === "2",
					"text-t3": title === "3",
					"text-t4": title === "4",
					"text-t5": title === "5",
					"text-t6": title === "6",
				},
				props.className,
			)}
		>
			{children}
		</Tag>
	);
};

export default Title;
