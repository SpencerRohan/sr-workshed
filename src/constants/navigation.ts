export const NAV_KEYS = {
	ARROW_LEFT: "ArrowLeft",
	ARROW_DOWN: "ArrowDown",
	ARROW_RIGHT: "ArrowRight",
	ARROW_UP: "ArrowUp",
	END: "End",
	ESCAPE: "Escape",
	HOME: "Home",
	PAGE_DOWN: "PageDown",
	PAGE_UP: "PageUp",
	RETURN: "Enter",
	SPACE: " ",
	TAB: "Tab",
} as const;

export const LOCK_KEYS = {
	CAPS_LOCK: "CapsLock",
	NUM_LOCK: "NumLock",
} as const;

export const FOCUS_DIRECTIONS = {
	FIRST: "first",
	LAST: "last",
	NEXT: "next",
	PREV: "previous",
} as const;
