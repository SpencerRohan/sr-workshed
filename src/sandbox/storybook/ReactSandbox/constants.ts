import React, { useEffect, useState } from "react";
import COMPONENTS from "../../../index";

export const REACT_LIVE_SAMPLE = `
  () => {
    const [buttonName, setButtonName] = useState("Click Here");
      
    return( 
      <div>
        <h2>Check out this awesome button!</h2>
        <button type="button" onClick={() =>setButtonName('Clicked')}>{buttonName}</button>
      </div>
    )
  }
`;
export const REACT_LIVE_SCOPE = { ...COMPONENTS, React, useEffect, useState };
