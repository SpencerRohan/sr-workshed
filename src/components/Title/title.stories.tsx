import getSelectObject from "@/utils/storybook/get-select-object";
import type { Meta, StoryObj } from "@storybook/react";
import { TITLE_HEADINGS, TITLE_VARIANTS } from "./constants";
import TitleComponent, { TitleProps } from "./index";

const meta = {
	title: "Components/Title",
	component: TitleComponent,
	tags: ["autodocs"],
	args: {
		children: "This is a heading rendered through the Title component",
	},
	argTypes: {
		as: {
			...getSelectObject(TITLE_HEADINGS),
			table: { type: { summary: Object.values(TITLE_HEADINGS).join(" | ") } },
		},
		title: {
			...getSelectObject(TITLE_VARIANTS),
			table: { type: { summary: Object.values(TITLE_VARIANTS).join(" | ") } },
		},
	},
} satisfies Meta<TitleProps>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Title: Story = {
	render: ({ children, ...args }) => (
		<TitleComponent {...args}>{children}</TitleComponent>
	),
};

// const Template = (args) => {
//   const { as, title, content, fontWeight } = args;
//   const weight = fontWeight ? `font-${fontWeight}` : "";

//   return (
//     <Title as={as} title="t"tle} className={weight}>
//       {content || `This is a heading ${as} styled as a Title ${title}`}
//     </Title>
//   );
// };

// const TailwindTemplate = (args) => {
//   const { fontWeight } = args;
//   const weight = fontWeight ? `font-${fontWeight}` : "";

//   return (
//     <div>
//       <h1 className={`text-t1 md:text-t1-md lg:text-t1-lg ${weight}`}>
//         Heading h1 | Title 1
//       </h1>
//       <h2 className={`text-t2 ${weight}`}>Heading h2 | Title 2</h2>
//       <h3 className={`text-t3 ${weight}`}>Heading h3 | Title 3</h3>
//       <h4 className={`text-t4 ${weight}`}>Heading h4 | Title 4</h4>
//       <h5 className={`text-t5 ${weight}`}>Heading h5 | Title 5</h5>
//       <h6 className={`text-t6 ${weight}`}>Heading h6 | Title 6</h6>
//     </div>
//   );
// };

// const Title = (args) => {
//   const { fontWeight } = args;
//   const weight = fontWeight ? `font-${fontWeight}` : "";

//   return (
//     <div>
//       <Title as="h1" title="1" className={weight}>
//         This is a Title 1
//       </Title>
//       <Title as="h1" title="2" className={weight}>
//         This is a Title 2
//       </Title>
//       <Title as="h1" title="3" className={weight}>
//         This is a Title 3
//       </Title>
//       <Title as="h1" title="4" className={weight}>
//         This is a Title 4
//       </Title>
//       <Title as="h1" title="5" className={weight}>
//         This is a Title 5
//       </Title>
//       <Title as="h1" title="6" className={weight}>
//         This is a Title 6
//       </Title>
//     </div>
//   );
// };

// export const Default = Template.bind({});
// export const Guide = GuideTemplate.bind({});
// export const WithTailwind = TailwindTemplate.bind({});
