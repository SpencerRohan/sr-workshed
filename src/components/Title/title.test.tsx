import conditionalDescribe from "@/utils/testing-library/conditional-describe";
import { render, screen } from "@testing-library/react";
import { TITLE_HEADINGS, TITLE_VARIANTS } from "./constants";

import Title from "./index";

const TITLE_TEST_ID = "title";
const HeadingCollection = () => (
	<>
		<Title>Heading Default</Title>
		{Object.values(TITLE_VARIANTS).map((variant, index) => {
			return (
				<Title
					as={Object.values(TITLE_HEADINGS)[index]}
					title={variant}
					key={index}
				>
					{`Heading h${index + 1} / Title ${index + 1}`}
				</Title>
			);
		})}
		<Title as="h1" title="2">
			Heading h1 / Title 2
		</Title>
		<Title
			as="h2"
			title="4"
			className="text-harris-dblue-400 text-center"
			aria-label="Title with additional settings"
		>
			<span>Title w/additional settings</span>
		</Title>
	</>
);

describe("Title", () => {
	describe("renders", () => {
		it("the proper default heading", () => {
			render(<Title data-testid={TITLE_TEST_ID}>Heading h1 / Title 1</Title>);
			const subject = screen.getByTestId(TITLE_TEST_ID);
			expect(subject).toBeInTheDocument();
			expect(subject.tagName).toEqual("H1");
		});

		it("the proper heading", () => {
			render(
				<Title as="h2" title="4" data-testid={TITLE_TEST_ID}>
					Heading h2 / Title 4
				</Title>,
			);
			const subject = screen.getByTestId(TITLE_TEST_ID);
			expect(subject).toBeInTheDocument();
			expect(subject.tagName).toEqual("H2");
		});

		it("the proper class", () => {
			render(
				<Title
					as="h2"
					title="4"
					data-testid={TITLE_TEST_ID}
					className="text-h6"
				>
					Heading h2 / Title 4
				</Title>,
			);
			const subject = screen.getByTestId(TITLE_TEST_ID);
			expect(subject).toBeInTheDocument();
			expect(subject).toHaveClass("text-h6 md:text-t4-md lg:text-t4-lg");
		});

		conditionalDescribe("snapshot", () => {
			it("matches", () => {
				const { asFragment } = render(<HeadingCollection />);
				expect(asFragment()).toMatchSnapshot();
			});
		});
	});
});
