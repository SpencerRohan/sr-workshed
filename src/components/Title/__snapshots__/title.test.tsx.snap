// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title renders snapshot matches 1`] = `
<DocumentFragment>
  <h1
    class="text-t1 md:text-t1-md lg:text-t1-lg"
  >
    Heading Default
  </h1>
  <h1
    class="text-t1 md:text-t1-md lg:text-t1-lg"
  >
    Heading h1 / Title 1
  </h1>
  <h2
    class="text-t2"
  >
    Heading h2 / Title 2
  </h2>
  <h3
    class="text-t3"
  >
    Heading h3 / Title 3
  </h3>
  <h4
    class="text-t4"
  >
    Heading h4 / Title 4
  </h4>
  <h5
    class="text-t5"
  >
    Heading h5 / Title 5
  </h5>
  <h6
    class="text-t6"
  >
    Heading h6 / Title 6
  </h6>
  <h1
    class="text-t2"
  >
    Heading h1 / Title 2
  </h1>
  <h2
    aria-label="Title with additional settings"
    class="text-t4 text-[pink] text-center"
  >
    <span>
      Title w/additional settings
    </span>
  </h2>
</DocumentFragment>
`;
